components/supervisor/pkg/supervisor/tasks.go

```diff
@@ -0,0 +1,412 @@
+		tasks:             make(map[string]*task),
+		subscriptions:     make(map[*tasksSubscription]struct{}),
+		ready:             make(chan struct{}),
+	}
+}
+
+func (tm *tasksManager) getStatus() []*api.TaskStatus {
+	tm.mu.RLock()
+	defer tm.mu.RUnlock()
+
+	i := 0
+	status := make([]*api.TaskStatus, len(tm.tasks))
+	for _, task := range tm.tasks {
+		status[i] = &task.TaskStatus
+		i++
+	}
+	return status
+}
+
+func (tm *tasksManager) updateState(doUpdate func() *task) {
+	tm.mu.Lock()
+	defer tm.mu.Unlock()
+
+	updated := doUpdate()
+	if updated == nil {
+		return
+	}
+	updates := make([]*api.TaskStatus, 1)
+	updates[0] = &updated.TaskStatus
+	for sub := range tm.subscriptions {
+		select {
+		case sub.updates <- updates:
+		default:
+			log.Warn("cannot to push tasks update to a subscriber")
+		}
+	}
+}
+
+func (tm *tasksManager) setTaskState(t *task, newState api.TaskState) {
+	tm.updateState(func() *task {
+		if t.State == newState {
+			return nil
+		}
+		t.State = newState
+		return t
+	})
+}
+
+func (tm *tasksManager) init(ctx context.Context) *runContext {
+	defer close(tm.ready)
+
+	tasks, err := tm.config.getGitpodTasks()
+	if err != nil {
+		log.WithError(err).Fatal()
+		return nil
+	}
+	if tasks == nil {
+		log.Info("no gitpod tasks found")
+		return nil
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil
+	case <-tm.inWorkspaceHelper.ContentReady():
+	}
+
+	contentSource, _ := tm.inWorkspaceHelper.ContentSource()
+	headless := tm.config.GitpodHeadless != nil && *tm.config.GitpodHeadless == "true"
+	runContext := &runContext{
+		contentSource: contentSource,
+		headless:      headless,
+		tasks:         make([]*task, 0),
+	}
+
+	for i, config := range *tasks {
+		id := strconv.Itoa(i)
+		presentation := &api.TaskPresentation{}
+		if config.Name != nil {
+			presentation.Name = *config.Name
+		} else {
+			presentation.Name = tm.terminalService.DefaultWorkdir
+		}
+		if config.OpenIn != nil {
+			presentation.OpenIn = *config.OpenIn
+		}
+		if config.OpenMode != nil {
+			presentation.OpenMode = *config.OpenMode
+		}
+		task := &task{
+			TaskStatus: api.TaskStatus{
+				Id:           id,
+				State:        api.TaskState_opening,
+				Presentation: presentation,
+			},
+			config: config,
+		}
+		task.command = task.getCommand(runContext)
+		if task.command == "" {
+			task.State = api.TaskState_closed
+		} else {
+			runContext.tasks = append(runContext.tasks, task)
+		}
+		tm.tasks[id] = task
+	}
+	return runContext
+}
+
+func (tm *tasksManager) Run(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	runContext := tm.init(ctx)
+	if runContext == nil {
+		return
+	}
+	if len(runContext.tasks) == 0 {
+		log.Info("no gitpod tasks to run")
+		return
+	}
+
+	for _, t := range runContext.tasks {
+		taskLog := log.WithField("command", t.command)
+		taskLog.Info("starting a task terminal...")
+		openRequest := &api.OpenTerminalRequest{}
+		if t.config.Env != nil {
+			openRequest.Env = *t.config.Env
+		} else {
```

Comment:
I don't think the else branch is necessary. If you don't need an env, just don't set it.